src/cdsetool/download.py

```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
+                
+    validity_check(Path(result_path),temp_path,feature,options,try_number)
     return feature.get("id")
 
+def validity_check(path: Path,temp_path: Path,product_info, verify_checksum: bool = True,options={},try_number = 0):
+    size = temp_path.stat().st_size
+    content_length = product_info.get("properties").get("services").get("download").get("size")
+    if size > content_length:
+            """Dowloaded more than the size of the file"""
+            temp_path.unlink()
+            raise(f"Content length mismatch for {path.name}")
+    elif size == content_length:
+        # Check integrity with MD5 checksum
+        if verify_checksum:
+            if _checksum_compare(temp_path, product_info):
+                shutil.move(temp_path, path)
+            else:
+                temp_path.unlink()
+                raise(f"Checksum mismatch for file {path.name}")
+    else:
+        """Partial dowload call back download feature"""
+        download_feature(product_info,path,options,try_number+1)
+
+def _checksum_compare(temp_path, product_info,block_size=2**13):
+    """Compare a given MD5 checksum with one calculated from a file."""
+    checksum = None
+    algo = None
+    filename = product_info.get("properties").get("title")
+    if "Checksum" in product_info:
+        checksum_list = product_info.get("Checksum")
+        if len(checksum_list) >0:
+            for checksum_dict in checksum_list:
+                if "Algorithm" in checksum_dict:
+                    algo = checksum_dict['Algorithm']
+                    if algo == "sha3-256":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.sha3_256() 
+                    elif algo == "MD5":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.md5()
+                        break
+                    elif algo == "BLAKE3":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.blake2b() # TODO ?
+                else:
+                    raise InvalidChecksumError("No checksum information found in product information. No Algorithm provided provided")
+        else:
+                    raise InvalidChecksumError("No checksum information found in product information. The Checksum list is empty")
+    if checksum == None:
+        raise InvalidChecksumError("No checksum algo provided is supported.")
+    file_path = temp_path
+    file_size = file_path.stat().st_size
+    # with _get_monitor({}).status() as status:
+    #     status.set_filename(f"{algo} - {filename}")
+    #     status.set_filesize(file_size)
+    with open(file_path, "rb") as f:
+        while True:
+            block_data = f.read(block_size)
+            if not block_data:
+                break
+            algo.update(block_data)
+            # status.add_progress(block_data)
+    return algo.hexdigest().lower() == checksum.lower()
```

Comment:
If you move this part into the download loop, you can avoid having to read the file again. You can essentially halve the amount of disk IO for each file.

Heres a simplified example
```python3
        checksum = hashlib.md5() # select the correct algo
        with open(temp_path, "wb") as file:
            for chunk in response.iter_content(chunk_size=1024 * 1024 * 5):
                if chunk:
                    file.write(chunk)
                    checksum.update(chunk)
                    status.add_progress(len(chunk))
        assert checksum.digest() == the_odata_checksum

Reply:
Here I am insecure would this work for partial file as well? (we shoul add the part that was already downloaded before to the checksum in case). In this case this part I did't implenet for this reason.